libs/ledgerjs/packages/hw-transport-react-native-ble/android/src/main/java/com/hwtransportreactnativeble/HwTransportReactNativeBleModule.kt

```diff
@@ -0,0 +1,187 @@
+                    putArray("serviceUUIDs", Arguments.createArray().apply{pushString(device.serviceId)})
+                })
+            }
+            // We hit this callback whenever a device is seen or goes away, this means we effectively
+            // replace the list instead of emitting new events one by one. This solves the
+            val event = Arguments.createMap().apply {
+                putString("event", "replace")
+                putString("type", "replace")
+                putMap("data", Arguments.createMap().apply {
+                    putArray("devices", devices)
+                })
+            }
+            eventEmitter.dispatch(event)
+        }
+        promise.resolve(1)
+    }
+
+    @ReactMethod
+    fun addListener(eventName: String) {
+        // Set up any upstream listeners or background tasks as necessary
+    }
+
+    @ReactMethod
+    fun removeListeners(count: Int) {
+        // Remove upstream listeners, stop unnecessary background tasks
+    }
+
+    @ReactMethod
+    fun stop(promise: Promise) {
+        Timber.d("$tag: \t stop scanning")
+        bleManager.stopScanning()
+        promise.resolve(1)
+    }
+
+    @ReactMethod
+    fun connect(uuid: String, promise: Promise) {
+        Timber.d("$tag: \t Attempt to connect to $uuid")
+        var consumed = false
+        bleManager.connect(
+            address = uuid,
+            onConnectSuccess = {
+                retriesLeft = 1
+                Timber.d("$tag: \t connection success")
+                if (!consumed) {
+                    promise.resolve(it.id)
+                    consumed = true
+                } else {
+                    print("already consumed")
+                }
+            },
+            onConnectError = {
+                if (it === "Device connection lost") {
+                    // We shouldn't have `Device connection lost` here, we should only get a failure
+                    // to connect.
+                    Timber.d("$tag: \t connection failure ignored, trying again")
+                    if (retriesLeft > 0) {
+                        retriesLeft -= 1
+                        connect(uuid, promise)
+                    }
+                } else if (!consumed) {
+                    Timber.d("$tag: \t connection failure")
+                    promise.reject("connectError", Exception(it))
+                    consumed = true
+                } else {
+                    print("already consumed")
+                }
+            })
+    }
+
+    @ReactMethod
+    fun onAppStateChange(awake: Boolean) {
+        Timber.d("$tag: \t onAppStateChange triggered")
+        eventEmitter.onAppStateChange(awake)
+    }
+
+    private var pendingEvent: Timer? = null
+    @ReactMethod
+    fun disconnect(promise: Promise) {
+        /// Prevent race condition between organic disconnect (allow open app) and explicit disconnection below.
+        pendingEvent = Timer()
```

Comment:
Wouldn't this run this task periodically? It's a hack to get past a bug in the native android transport and it's supposed to be a one shot delay, introducing the period seems like it would run it every 3 seconds 🤔

Reply:
Ok, I was not sure what we wanted to do with it. You can rewrite this way, but it doesn't come with tremendous improvement:
```kotlin
pendingEvent?.schedule(disconnectDelay) {
    queue?.stop()
    if (!bleManager.isConnected) {
        promise.resolve(true)
    } else {
        bleManager.disconnect {
            Timber.d("$tag: \t disconnected")
            promise.resolve(true)
        }
    }
}
```